### PR TITLE
feat: add support for protocol negotiation fallback with custom arrays

### PR DESCRIPTION
**Pull request by anubhav756**
## Description
This PR adds support for specifying a custom array of supported MCP protocol versions during client initialization. This enables robust protocol negotiation and fallback strategies when interacting with different versions of the Toolbox server.

## Changes
- Implemented `core.WithSupportedProtocols([]core.Protocol{})` in the client options to handle custom array fallbacks.
- Added explicit aliases for `core.MCPLatest` and `core.MCPDraft` to the core constants.
- Refactored the E2E tests to use `serverURLStable` and `serverURLDraft` constants.
- Added explicit integration tests for establishing connections with `core.MCPLatest`, and verifying that the array fallback `[]core.Protocol{core.MCPLatest, core.MCPDraft}` successfully negotiates the correct protocol version.
- Updated the SDK `README.md` to document the new custom array examples for protocol negotiation.

### CODE REVIEW

File: core/client_test.go [hunk 1]
```diff
@@ -975,6 +975,157 @@ func TestExecuteWithFallback_EdgeCases(t *testing.T) {
...
+			_ = json.Unmarshal(body, &req)
+
+			var result any
+			switch req.Method {
+			case "initialize":
+				paramsMap, _ := req.Params.(map[string]any)
+				reqVersion, _ := paramsMap["protocolVersion"].(string)
+				if reqVersion == "2025-11-25" {
+					result = map[string]any{
+						"protocolVersion": "2025-03-26",
```
Comment (Yuan325) [inline core/client_test.go]: if returning a result, the protocol version would be the ones that the user is requesting (e.g. if reqVersion is 2025-11-25, the protocolVersion value will be 2025-11-25).

Else, it will return with invalid protocol version. (with or without the supported array).

Please double check the logic of the tests.

Comment (Yuan325) [reply core/client_test.go]: Please also double check the remaining 2 tests. 

It might be easier to mock the exact response (since we are only running 1 test case on this test server), rather than having a whole switch case to determine what to respond.

File: core/options_test.go [hunk 2]
```diff
@@ -635,3 +635,42 @@ func TestNewToolConfig(t *testing.T) {
 		t.Errorf("Expected Strict to be false, but got %t", config.Strict)
 	}
 }
+
+func TestWithSupportedProtocols(t *testing.T) {
```
Comment (Yuan325) [inline core/options_test.go:640]: can we also test one with `protocolSet`, and one that hits the error `WithSupportedProtocols: none of the provided protocols are supported`.

File: core/options_test.go [hunk 3]
```diff
@@ -635,3 +635,42 @@ func TestNewToolConfig(t *testing.T) {
...
+	})
+
+	t.Run("Invalid protocol element error", func(t *testing.T) {
+		client := newTestClient()
+		opt := WithSupportedProtocols([]Protocol{MCPv20251125, Protocol("invalid-version")})
+		err := opt(client)
+		if err == nil {
+			t.Error("Expected error for invalid protocol version element, got nil")
+		}
+		if !strings.Contains(err.Error(), "invalid protocol version 'invalid-version'") {
```
Comment (Yuan325) [inline core/options_test.go]: Might want to check the exact error message rather than `strings.Contains()`, to make sure that the allVersion array is parsed correctly.

**Review by Yuan325 — CHANGES_REQUESTED** [review]

File: core/client_test.go (hunk 1 above)
Comment (anubhav756) [reply core/client_test.go]: Makes sense! Thanks for the suggestion!

Refactored mock handlers in all 3 fallback unit tests to be compliant and simplified!

File: core/options_test.go (hunk 2 above)
Comment (anubhav756) [reply core/options_test.go:640]: Added the test case for `protocolSet` under `TestWithSupportedProtocols/Protocol_already_set_error`.

Regarding testing the `"WithSupportedProtocols: none of the provided protocols are supported"` error, I double-checked and realised that check (`if len(sorted) == 0`) is unreachable.

The loop preceding it (`for _, p := range protocols`) validates that every element in `protocols` is contained in `allVersions` and returns an `"invalid protocol version"` error if any item is unsupported. So by line `48` all input protocols are guaranteed to be valid supported versions, so sorted will always contain at least 1 element and `len(sorted) == 0` can never be to `true`.

I have removed the unreachable block. Thanks for bringing this up though!

File: core/options_test.go (hunk 3 above)
Comment (anubhav756) [reply core/options_test.go]: Thanks for the suggestion! Done.

File: core/transport/mcp/v20260618/mcp.go [hunk 4]
```diff
@@ -174,15 +180,30 @@ func (t *McpTransport) InvokeTool(ctx context.Context, toolName string, payload
 	return t.ProcessToolResultContent(baseContent), nil
 }
 
+var supportedVersionsPriority = []string{
```
Comment (Yuan325) [inline core/transport/mcp/v20260618/mcp.go:183]: Might want to add ProtocolVersion here? I think this might be a conflict from the other PR? Please double check and make sure those conflicts are merged correctly~

**Review by Yuan325 — APPROVED** [review]

File: core/transport/mcp/v20260618/mcp.go (hunk 4 above)
Comment (anubhav756) [reply core/transport/mcp/v20260618/mcp.go:183]: Looks like this was missing due to a merge conflict with renaming of files. This is already added in https://github.com/googleapis/mcp-toolbox-sdk-go/pull/309
